### PR TITLE
Accept constant function as arguments of within operator

### PR DESCRIPTION
**Pull request by eugenels**
There is dependency on https://github.com/questdb/questdb/pull/1265
Replaces https://github.com/questdb/questdb/pull/1276

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 80 (85.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|27|39|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|25|25|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 68 / 80 (85.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|27|39|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|25|25|100.00%|

File: core/src/main/java/io/questdb/cairo/GeoHashes.java [hunk 1]
```diff
@@ -187,7 +187,7 @@ private static long appendChar(long geohash, char c) throws NumericException {
         }
         throw NumericException.INSTANCE;
     }
-
+/*
     public static void fromStringToBits(final CharSequenceHashSet prefixes, int columnType, final DirectLongList prefixesBits) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/GeoHashes.java:191]: better to clean it up

File: core/src/main/java/io/questdb/cairo/ColumnType.java [hunk 2]
```diff
@@ -189,7 +189,6 @@ private ColumnType() {
     }
 
     public static int geohashWithPrecision(int bits) {
-        assert bits > 0;
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/ColumnType.java:192]: Asserts are not executed at prod environment, there is usually no reason to remove them unless they are meaningless. Any reason to remove this one?

File: core/src/main/java/io/questdb/cairo/GeoHashes.java [hunk 3]
```diff
@@ -217,7 +217,45 @@ public static void fromStringToBits(final CharSequenceHashSet prefixes, int colu
             }
         }
     }
+*/
+    public static void fromStringToBits(final CharSequenceHashSet prefixes, int columnType, final LongList prefixesBits) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/cairo/GeoHashes.java:221]: Is this method used anywhere other when tests? You can perhaps move it to the test code

File: core/src/main/java/io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java [hunk 4]
```diff
@@ -44,23 +42,22 @@ public LatestByAllIndexedFilteredRecordCursorFactory(
...
+            @NotNull LongList prefixes
     ) {
         super(metadata, dataFrameCursorFactory, configuration);
+        this.prefixes = new DirectLongList(Math.max(2, prefixes.size()));
 
-        this.prefixes = new DirectLongList(64);
-        if (hashColumnIndex > -1 && ColumnType.isGeoHash(hashColumnType)) {
-            GeoHashes.fromStringToBits(prefixes, hashColumnType, this.prefixes);
+        // copy into owned direct memory
+        for (int i = 0; i < prefixes.size(); i++) {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java:53]: Perhaps you can replace loop with `this.prefixes.add(prefixes);`

File: core/src/main/java/io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java [hunk 5]
```diff
@@ -44,23 +42,22 @@ public LatestByAllIndexedFilteredRecordCursorFactory(
             @NotNull RecordMetadata metadata,
             @NotNull DataFrameCursorFactory dataFrameCursorFactory,
             int columnIndex,
-            int hashColumnIndex,
-            int hashColumnType,
             @Nullable Function filter,
             @NotNull IntList columnIndexes,
-            @NotNull CharSequenceHashSet prefixes
+            @NotNull LongList prefixes
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java:47]: If this were DirectLongList then copying would be a bit faster.

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 6]
```diff
@@ -1158,6 +1180,42 @@ private boolean removeWithin(AliasTranslator translator, ExpressionNode node, Re
...
+            GeoHashes.addNormalizedGeoPrefix(unquote(inArg.token), columnType, prefixes);
+        }
+    }
+
+    private boolean isNull(ExpressionNode node) {
+        return node == null || isNullKeyword(node.token);
+    }
+
+    private boolean isGeoHashConstFunction(Function fn) {
+        return (fn instanceof GeoHashFunction) && (fn.isConstant() || fn.isRuntimeConstant());
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1216]: You cannot get value of Runtime Constant at compilation time. Example of Runtime Constant function is now() which is different every execution even when CursorFactory is reused

File: core/src/main/java/io/questdb/cairo/ColumnType.java (hunk 2 above)
Comment (eugenels) [reply core/src/main/java/io/questdb/cairo/ColumnType.java:192]: It should be assert bits > 0 && bits < 61;
But there is a lot of code with this check explicit

File: core/src/main/java/io/questdb/cairo/GeoHashes.java (hunk 1 above)
Comment (eugenels) [reply core/src/main/java/io/questdb/cairo/GeoHashes.java]: fixed

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 74 (83.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|27|39|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|19|19|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 62 / 74 (83.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|27|39|69.23%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|19|19|100.00%|

**Review by marregui — COMMENTED** [review]
LGTM

I would add explicit exemplifying tests for `addNormalizedGeoPrefix` in GeoHashesTest.
Geohashes literal PR has been merged.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 89 (85.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|41|54|75.93%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|19|19|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 89 (85.39%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|41|54|75.93%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|19|19|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 80 (95.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 80 (95.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 7]
```diff
@@ -1158,6 +1181,64 @@ private boolean removeWithin(AliasTranslator translator, ExpressionNode node, Re
...
+            LongList prefixes
+    ) throws SqlException, NumericException {
+        if(isNull(inArg)) {
+            throw SqlException.$(inArg.position, "GeoHash value expected");
+        }
+        if (isFunc(inArg)) {
+            Function f = functionParser.parseFunction(inArg, metadata, executionContext);
+            if (isGeoHashConstFunction(f)) {
+                final int fnType = f.getType();
+                final long hash = f.getGeoHashLong(null);
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1199]: Functions will / should not allow to call wrong type method. Some geohash resulting functions like CAST(String, Geohash) will throw assertion if you do `getGeoHashLong` on  say 7bit geohash type

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 8]
```diff
@@ -1158,6 +1181,64 @@ private boolean removeWithin(AliasTranslator translator, ExpressionNode node, Re
...
+            final boolean isCharsPrefix = len > 1 && inArg.token.charAt(0) == '#';
+
+            if (!(isConstant && (isBitsPrefix || isCharsPrefix))) {
+                throw SqlException.$(inArg.position, "GeoHash literal expected");
+            }
+
+            final int type;
+            final long hash;
+            if (isBitsPrefix) {
+                type = ColumnType.geohashWithPrecision(inArg.token.length() - 2);
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1217]: Can length be > 62 here? It will result in invalid type

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 80 (95.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Review by ideoma — CHANGES_REQUESTED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 76 / 80 (95.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java (hunk 8 above)
Comment (eugenels) [reply core/src/main/java/io/questdb/griffin/WhereClauseParser.java:1217]: ExpressionParser will throw exception on invalid bits literal before this code.
addNormalizedGeoPrefix does prefix/column precision mismatch check later

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 84 (95.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashConstant.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 84 (95.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|50|54|92.59%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashConstant.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Review by ideoma — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 84 / 86 (97.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|54|56|96.43%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashConstant.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 84 / 86 (97.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/WhereClauseParser.java|54|56|96.43%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedFilteredRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/LatestByAllIndexedRecordCursor.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/constants/GeoHashConstant.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnType.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/GeoHashes.java|9|9|100.00%|

**Comment by eugenels** [comment]
replaced by: https://github.com/questdb/questdb/pull/1307

**Pull request closed without merging**